web/src/message_list.js

```diff
@@ -68,18 +68,39 @@ export class MessageList {
         // the user. Possibly this can be unified in some nice way.
         this.reading_prevented = false;
 
-        // Whether this message list's is preserved in the DOM even
+        return this;
+    }
+
+    should_preserve_current_rendered_state() {
+        // Whether this message list is preserved in the DOM even
         // when viewing other views -- a valuable optimization for
         // fast toggling between the combined feed and other views,
         // which we enable only when that is the user's home view.
         //
         // This is intentionally not live-updated when web_home_view
         // changes, since it's easier to reason about if this
         // optimization is active or not for an entire session.
-        this.preserve_rendered_state =
-            user_settings.web_home_view === "all_messages" && !this.narrowed;
+        if (user_settings.web_home_view !== "all_messages" || this.narrowed) {
+            return false;
+        }
 
-        return this;
+        // Since our natural tendency is to narrow user to `first_unread` message,
+        // we don't preserve the rendered state of this list if first unread message
+        // is not available to us. Otherwise, this leads to confusion when we are
+        // restoring the rendered list but our first unread message is not available
+        // and fetching it from the server could lead to non-contiguous messages history.
+        //
+        // NOTE: Non-contiguous message history can still happen in the opposite situation
+        // where user is narrowing to a message id which is not present in the rendered list.
+        // In this case, we create a new list and if the new fetched history contains first
+        // unread message, we preserver this list and discard others with the same filter.
+        const first_unread_message = this.get(this.data.first_unread_message_id());
+        if (!first_unread_message?.unread) {
+            // If we have found the newest message, we can preserve the rendered state.
+            return this.data.fetch_status.has_found_newest();
+        }
```

Comment:
I don't understand why wither the `first_unread_message` calculation/check is relevant or why `this.data.fetch_status.has_found_newest` is the right check -- I feel like I'm missing something basic about the strategy here.

Reply:
Modified the comment:

```

        // If we click on a narrow, we go the first unread message.
        // If first unread message is not available in a cached message list,
        // we render by selecting the `message_list.last()` message. 
        // This is incorrect unless we have found the newest message.
        //
        // So, we don't preserve the rendered state of this list if first unread message
        // is not available to us. Otherwise, this leads to confusion when we are
        // restoring the rendered list but our first unread message is not available
        // and fetching it from the server could lead to non-contiguous messages history.
        //
        // NOTE: Non-contiguous message history can still happen in the opposite situation
        // where user is narrowing to a message id which is not present in the rendered list.
        // In this case, we create a new list and if the new fetched history contains first
        // unread message, we preserver this list and discard others with the same filter.
```

Added more comments here:


```js
    first_unread_message_id(): number | undefined {
        // NOTE: This function returns the first unread that was fetched and is not
        // necessarily the first unread for the narrow. There could be more unread messages.
        // See https://github.com/zulip/zulip/pull/30008#discussion_r1597279862 for how
        // ideas on how to possibly improve this.
        // before `first_unread` calculated below that we haven't fetched yet.
        const first_unread = this._items.find((message) => unread.message_unread(message));

        if (first_unread) {
            return first_unread.id;
        }

        // If no unread, return the bottom message
        // NOTE: This is only valid if we have found the latest message for the narrow as
        // there could be more message that we haven't fetched yet.
        return this.last()?.id;
```